[Discover][DocViewer] Convert EuiTable to EuiDataGrid. Enable up to 500 fields per page.

jughosta opened the pull request:
- Closes https://github.com/elastic/kibana/issues/174745

## Summary

This PR converts Doc Viewer table into EuiDataGrid to use its actions functionality.

<img width="703" alt="Screenshot 2024-05-17 at 20 18 44" src="https://github.com/elastic/kibana/assets/1415710/10d8a7b0-8fe1-4908-a11d-5fd374eed4c3">
<img width="577" alt="Screenshot 2024-05-17 at 18 17 49" src="https://github.com/elastic/kibana/assets/1415710/7e6f05ce-9690-48ab-84c0-f8776e360f83">
<img width="490" alt="Screenshot 2024-05-17 at 18 18 05" src="https://github.com/elastic/kibana/assets/1415710/b36c64de-419d-425c-9890-8bc346059c1a">
<img width="871" alt="Screenshot 2024-05-22 at 15 22 31" src="https://github.com/elastic/kibana/assets/1415710/92c894f3-91f8-445c-b6fb-ba8842dd3b23">


## Testing

Some cases to check while testing:
- varios value formats
- legacy table vs data grid
- doc viewer flyout vs Single Document page

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

jughosta commented:
/ci

jughosta commented:
/ci

jughosta commented:
/ci

jughosta commented:
/ci

jughosta commented:
/ci

jughosta commented:
/ci

jughosta commented:
/ci

jughosta commented:
/ci

kertal reviewed (commented):
Looks pretty nice 👍 , one thing I noticed while testing, it seems field with custom label can be pinned but the icon is not displayed
<img width="534" alt="Discover_-_Elastic" src="https://github.com/elastic/kibana/assets/463851/fdaf4154-fae6-42e5-abd1-3b0d94b3882e">

jughosta commented:
> one thing I noticed while testing, it seems field with custom label can be pinned but the icon is not displayed

@kertal Good catch! I think we have this issue on main too. I can address it separately and backport.

jughosta commented:
/ci

jughosta commented:
/ci

andreadelrio reviewed (changes requested):
@jughosta looking good!  Here's a couple of comments:

- Can you flush the empty buttons in the EuiPopoverFooter to the left? That way, visually, we'll have equal padding to the section above the horizontal line.

<img width="423" alt="image" src="https://github.com/elastic/kibana/assets/4016496/7f88b66a-8520-4b56-b35b-9122436bf27a">

- I think this tooltip is unnecessary and it covers the floating cell actions. 
<img width="188" alt="image" src="https://github.com/elastic/kibana/assets/4016496/d0b8500a-96a2-45f6-8dce-d366f3f2e42c">

jughosta commented:
@andreadelrio Thanks for the review! I updated the PR.

jughosta commented:
@elasticmachine merge upstream

jughosta commented:
@elasticmachine merge upstream

andreadelrio commented on packages/kbn-unified-doc-viewer/src/components/field_name/field_name.tsx:
```diff
@@ -41,13 +50,22 @@ export function FieldName({
+      <EuiFlexGroup gutterSize="none" responsive={false} alignItems="center" direction="row" wrap>
+        {isPinned && (
+          <EuiFlexItem grow={false} className="kbnDocViewer__fieldPinned">
+            <EuiIcon type="pinFilled" />
```
```suggestion
            <EuiIcon type="pinFilled" />
```
```suggestion
            <EuiIcon type="pinFilled" size="s" />
```

andreadelrio reviewed (approved):
@jughosta Design changes LGTM! Just left a comment about reducing the size of the pin icon. Thanks for working on this, it's the perfect addition to the push flyout.

davismcphee reviewed (commented):
The code changes look great! It feels like a significant improvement over the old table too. I did some quick local testing and so far so good, but I'll follow up with some more testing tomorrow and leave another review.

jughosta commented on packages/kbn-unified-doc-viewer/src/components/field_name/field_name.tsx:
```diff
@@ -41,13 +50,22 @@ export function FieldName({
+      <EuiFlexGroup gutterSize="none" responsive={false} alignItems="center" direction="row" wrap>
+        {isPinned && (
+          <EuiFlexItem grow={false} className="kbnDocViewer__fieldPinned">
+            <EuiIcon type="pinFilled" />
```
@davismcphee As it might obscure the cell actions, I removed it per https://github.com/elastic/kibana/pull/175787#pullrequestreview-2075161962

jughosta replied:
Added it as `aria-label` https://github.com/elastic/kibana/pull/175787/commits/08ce88d4f89ecdc4bbaaa8d980084bcf3075b538

jughosta commented on src/plugins/unified_doc_viewer/public/components/doc_viewer_table/table_cell_actions.tsx:
```diff
@@ -6,125 +6,212 @@
-  icon: string;
-  onClick: () => void;
+export function isFilterInOutPairDisabled(row: TableRow): boolean {
+  if (!row) {
+    return false;
+  }
```
Added a comment in https://github.com/elastic/kibana/pull/175787/commits/62354826ec195dfdc14e999d646aeb6bbb54b5fe

jughosta replied:
Updated via https://github.com/elastic/kibana/pull/175787/commits/b2a79ef2052300a47e6368fedddcf9060d2a2db4

jughosta commented:
@elasticmachine merge upstream

davismcphee reviewed (commented):
Works great overall! Just noticed a couple of things while testing that we should look into.

When resizing the window after the flyout is opened, the data grid doesn't adjust its height to match. This also happens in single document view:

https://github.com/elastic/kibana/assets/25592674/d6e43095-947b-48d9-8b98-6f3e53ea9502

When using the keyboard to navigating left or right through the data grid cells, the selected doc changes too:

https://github.com/elastic/kibana/assets/25592674/309dde78-5d74-4294-afc1-0c0723882d8c

This one is very minor, but I noticed theres now a bit of extra space below the grid in single document view:
<img width="1998" alt="padding" src="https://github.com/elastic/kibana/assets/25592674/832a6c41-90dd-40e7-ae7d-fb2266632557">

stratoula reviewed (commented):
I noticed something weird. 

When you are in the dataview mode with the flyout open and you navigate to ESQL, the flyout remains open and the document is wrong. You can understand this because you see the _id which is not available in ES|QL (unless you specifically request it with the metadata command.

![meow](https://github.com/elastic/kibana/assets/17003240/d22f3645-d3fb-49aa-94ea-92ed06ac1465)

Surprisingly the bug doesnt exist when I go to the dataview mode.

- Can we do something for fields with large names? In the previous implementation looked much better

<img width="205" alt="image" src="https://github.com/elastic/kibana/assets/17003240/0d597e22-c865-46f4-bc0d-ff7bce858290">

- Also pinning fields with large names seems a bit off
<img width="518" alt="image" src="https://github.com/elastic/kibana/assets/17003240/486191de-e564-40dd-b162-cc45aea332d6">

stratoula commented:
Also, this happens in main too, it is a bit more obvious here though as the text has changed from "Toggle column" to "Add column". The timestamp field is already in the table so adding is wrong here. This happens in both modes, no big deal but is confusing. In the table we don't allow the @timestamp to be removed. Maybe we should disable this action from the flyout too? Wdyt?

stratoula commented on src/plugins/esql_datagrid/public/row_viewer.test.tsx:
```diff
@@ -106,14 +106,14 @@ describe('RowViewer', () => {
     } as unknown as DataTableRecord);
     await waitFor(() => {
-      expect(screen.getByTestId('esqlTableRowNavigation')).toBeInTheDocument();
+      expect(screen.getByTestId('dscDocNavigation')).toBeInTheDocument();
```
Can we make the test sunj here more generic and not discover specific? (As it is a shared component now)

jughosta replied:
Updated in https://github.com/elastic/kibana/pull/175787/commits/dd680be08e7cf5ef7460a0545bfacefc4c607e6f

jughosta commented:
Hi @davismcphee and @stratoula,

Thanks for your feedback! Please let me know if I missed something. Here are updates since you last checked:

> When resizing the window after the flyout is opened, the data grid doesn't adjust its height to match. This also happens in single document view

Addressed via https://github.com/elastic/kibana/pull/175787/commits/b012bce6f021508dec0ec584573353be52ecd282

> When using the keyboard to navigating left or right through the data grid cells, the selected doc changes too:

Addressed via https://github.com/elastic/kibana/pull/175787/commits/3d114008cab049b8ad89d1cff7828a5cc3771622

> This one is very minor, but I noticed theres now a bit of extra space below the grid in single document view

Addressed via https://github.com/elastic/kibana/pull/175787/commits/b012bce6f021508dec0ec584573353be52ecd282

> When you are in the dataview mode with the flyout open and you navigate to ESQL, the flyout remains open and the document is wrong. 

Addressed via https://github.com/elastic/kibana/pull/175787/commits/b4b7b30d86f809802af13cf4ccc83546a0834087

> Can we do something for fields with large names?

Addressed via https://github.com/elastic/kibana/pull/175787/commits/440f171f5e374a6cacc416d803646ad9cffb64c0

> Also, this happens in main too, it is a bit more obvious here though as the text has changed from "Toggle column" to "Add column". The timestamp field is already in the table so adding is wrong here. This happens in both modes, no big deal but is confusing. In the table we don't allow the @timestamp to be removed. Maybe we should disable this action from the flyout too? Wdyt?

Updated the logic to be back like in main via https://github.com/elastic/kibana/pull/175787/commits/74c9eb599dd86894ae4b93e07f97429f8ba083c0 I think it's okay to allow toggling the time field in columns. We allow it from the sidebar too. It does not really affect the table view. Anyway, it can be enhanced separately.

stratoula commented:
>  I think it's okay to allow toggling the time field in columns. We allow it from the sidebar too. It does not really affect the table view. Anyway, it can be enhanced separately.

It can def enhanced separately, it also happens in main. I think the best UX is to hide the action. If we allow the action then it should affect the table. As it is right now is weird imho

stratoula commented:
I created an issue for the confusing behavior I found above here https://github.com/elastic/kibana/issues/185637

stratoula reviewed (approved):
This works great now, LGTM

davismcphee commented on src/plugins/unified_doc_viewer/public/index.tsx:
```diff
@@ -26,14 +25,11 @@ export const JsonCodeEditor = withSuspense<JsonCodeEditorProps>(
 export { useEsDocSearch } from './hooks';
+export { UnifiedDocViewer } from './components/lazy_doc_viewer';
+export {
+  UnifiedDocViewerFlyout,
```
I noticed the page load bundle increased `5.9KB` with this PR. Could we also lazy load `UnifiedDocViewerFlyout`?

davismcphee reviewed (approved):
Tested again, and the issues I noticed have now been resolved. Also thanks for migrating the doc viewer flyout to the plugin to be shared! LGTM 👍

>Updated the logic to be back like in main via https://github.com/elastic/kibana/commit/74c9eb599dd86894ae4b93e07f97429f8ba083c0 I think it's okay to allow toggling the time field in columns. We allow it from the sidebar too. It does not really affect the table view. Anyway, it can be enhanced separately.

Personally I preferred how you had it even with the timestamp issue. I think if we don't include these changes in this PR, we should at least bring them back when addressing #185637.
